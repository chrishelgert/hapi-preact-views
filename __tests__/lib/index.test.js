jest.mock('decache')

const path = require('path')
const Hapi = require('hapi')
const vision = require('vision')
const decache = require('decache')
const hapiPreactViews = require('../../lib/index')

describe('hapi-preact-views', () => {
  it('should return a compile-functions', () => {
    expect(hapiPreactViews.compile).toBeInstanceOf(Function)
  })

  describe('compile', () => {
    const context = { name: 'Test' }
    const layoutConfig = {
      layoutPath: path.join(__dirname, '..', '..', 'example', 'layouts'),
      layout: 'Default',
    }
    let server

    function createServer(compileOptions = {}) {
      return new Promise((resolve) => {
        server = new Hapi.Server()
        server.register(vision, () => {
          server.views({
            engines: { js: hapiPreactViews },
            relativeTo: path.resolve(__dirname, '..', '..', 'example'),
            path: 'views',
            compileOptions,
          })

          resolve()
        })
      })
    }

    describe('without layout', () => {
      it('should return the view', (done) => {
        createServer().then(() => {
          server.render('View', context, (err, output) => {
            expect(err).toBeNull()
            expect(output).toMatchSnapshot()

            done()
          })
        })
      })

      it('should clear the cache, when option.cache is false', (done) => {
        createServer({ cache: false }).then(() => {
          server.render('View', context, () => {
            expect(decache).toHaveBeenCalled()

            done()
          })
        })
      })
    })

    describe('with layout', () => {
      it('should render the view in the layout', (done) => {
        const config = Object.assign(layoutConfig, { cache: true })
        createServer(config).then(() => {
          server.render('View', context, (err, output) => {
            expect(err).toBeNull()
            expect(output).toMatchSnapshot()

            done()
          })
        })
      })

      it('should clear the cache, when option.cache is false', (done) => {
        const config = Object.assign(layoutConfig, { cache: false })
        createServer(config).then(() => {
          server.render('View', context, () => {
            expect(decache).toHaveBeenCalled()

            done()
          })
        })
      })
    })

    it('should call beforeRender, when it is in options', (done) => {
      const beforeRender = jest.fn()
      createServer({ beforeRender }).then(() => {
        server.render('View', context, () => {
          expect(beforeRender).toHaveBeenCalled()

          done()
        })
      })
    })
  })
})
