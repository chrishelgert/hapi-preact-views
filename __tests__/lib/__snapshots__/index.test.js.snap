exports[`hapi-preact-views compile with layout should render the view in the layout 1`] = `"<!DOCTYPE html><html lang=\"en\"><head><meta name=\"viewport\" content=\"with=device-with; initial-scale=1.0\"></head><body><section id=\"default-layout\"><h1>Hello Test</h1></section></body></html>"`;

exports[`hapi-preact-views compile without layout should return the view 1`] = `"<!DOCTYPE html><h1>Hello Test</h1>"`;
